src/bindings/jna/process/src/main/java/org/libelektra/process/ProcessApp.java

```diff
@@ -0,0 +1,55 @@
+package org.libelektra.process;
+
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import org.libelektra.Plugin;
+
+/** Implements the child side of the protocol used by the process plugin. */
+public class ProcessApp {
+
+  public static void main(String[] args) {
+    if (args.length != 1) {
+      System.exit(1);
+    }
+
+    Class<Plugin> pluginClass;
+    try {
+      Class<?> clazz = Class.forName(args[0]);
+      if (!Plugin.class.isAssignableFrom(clazz)) {
+        System.exit(1);
+        return;
+      }
+      //noinspection unchecked
+      pluginClass = (Class<Plugin>) clazz;
+    } catch (ClassNotFoundException e) {
+      System.exit(1);
+      return;
+    }
+
+    Plugin plugin;
+    try {
+      plugin = pluginClass.getConstructor().newInstance();
+    } catch (InstantiationException
+        | IllegalAccessException
+        | InvocationTargetException
+        | NoSuchMethodException e) {
+      System.exit(1);
+      return;
+    }
+
+    var protocol = new ProcessProtocol(plugin, System.in, System.out);
+
+    try {
+      if (!protocol.handshake()) {
+        System.exit(1);
```

Comment:
Since the handshake failed, the only real option here is to print something to `stderr`. However, that could be really confusing, because (under normal circumstances) we are not in the process that was called by the user.

Even if we print error messages to `stderr` (which could be really out of place depending on what the parent process is doing), in this specific case, I don't think we need one. If the handshake failed, there isn't much information we could provide. If the app was called by the `process` plugin and something went wrong, `process` will issue a "handshake failed" error. If the app was called in some other way, just exiting because the app is being misused seems fine.

Reply:
We could exit with 3 different error status? Then the plugin can wait for the child and knows the exit status for an error message.

Reply:
I'm not sure that would have much benefit. The three cases would be:
1. "Handshake failed": Without any extra information the separate exit code is useless. The plugin already knows that the handshake failed, because it initiated the handshake which the child process aborted.
2. An IOException: In this case, maybe a different error code could convey some information. The parent process probably also got some error from `read`/`write`ing to its end of the pipes.
3. Another error during the protocol execution: This mostly boils down to an encoding/decoding error of the dump format. This shouldn't happen and if it does it's a bug.

Additionally, in all 3 cases the user can't really do anything to solve the error. Not even the developer of the Java plugin could do anything here. All these errors are essentially bugs that need to be reported and fixed by us (*).

---

(*) I didn't use `ELEKTRA_SET_INTERNAL_ERROR` in the `process` plugin, because the errors are only internal errors, if the child app is also provided us (like with Java). If it's an external app, the developer of that app needs to fix the problem.

Reply:
> "Handshake failed": Without any extra information the separate exit code is useless. The plugin already knows that the handshake failed, because it initiated the handshake which the child process aborted.

It does not know if the child agrees about that the handshake failed. Even if this is very unlikely/impossible there is also not a good reason to use the same exit code?

> An IOException: In this case, maybe a different error code could convey some information. The parent process probably also got some error from read/writeing to its end of the pipes.

Good idea!

> Additionally, in all 3 cases the user can't really do anything to solve the error. Not even the developer of the Java plugin could do anything here. All these errors are essentially bugs that need to be reported and fixed by us (*).

Yes, so the error should be designed that it is helpful for us.

Reply:
> What do you mean "the child agrees"?

The child might have written the ACK to stdout but the parent process does not receive it (e.g. wrongly connected stdout).

> On a more technical note, it also requires quite extensive changes to the process plugin's code.

If it makes the code much more complicated then leave it. We will see if errors in the protocol occurs and if we will be able to debug it easily :wink: